discovery_engine_core/ai/rubert/src/config.rs

```diff
@@ -0,0 +1,103 @@
+use std::{
+    fs::File,
+    io::{BufRead, BufReader, Read},
+    marker::PhantomData,
+    path::Path,
+};
+
+use displaydoc::Display;
+use thiserror::Error;
+
+use crate::{model::BertModel, NonePooler};
+
+#[derive(Debug, Display, Error)]
+pub enum ConfigError {
+    /// The token size must be greater than two to allow for special tokens
+    TokenSize,
+    /// Failed to load a data file: {0}
+    DataFile(#[from] std::io::Error),
+}
+
+pub struct Config<'a, K, P> {
+    pub(crate) model_kind: PhantomData<K>,
+    pub(crate) vocab: Box<dyn BufRead + Send + 'a>,
+    pub(crate) model: Box<dyn Read + Send + 'a>,
+    pub(crate) accents: bool,
+    pub(crate) lowercase: bool,
+    pub(crate) token_size: usize,
+    pub(crate) pooler: P,
+}
+
+impl<'a, K: BertModel> Config<'a, K, NonePooler> {
+    pub fn from_readers(
+        vocab: Box<dyn BufRead + Send + 'a>,
+        model: Box<dyn Read + Send + 'a>,
+    ) -> Self {
+        Config {
+            model_kind: Default::default(),
+            vocab,
+            model,
+            accents: false,
+            lowercase: true,
+            token_size: 128,
+            pooler: NonePooler,
+        }
+    }
+
+    pub fn from_files(
+        vocab: impl AsRef<Path>,
+        model: impl AsRef<Path>,
+    ) -> Result<Self, ConfigError> {
+        let vocab = Box::new(BufReader::new(File::open(vocab)?));
+        let model = Box::new(BufReader::new(File::open(model)?));
+        Ok(Self::from_readers(vocab, model))
+    }
+}
+
+impl<'a, K: BertModel, P> Config<'a, K, P> {
+    /// Whether the tokenizer keeps accents.
+    ///
+    /// Defaults to `false`.
+    pub fn with_accents(mut self, accents: bool) -> Self {
+        self.accents = accents;
+        self
+    }
+
+    /// Whether the tokenizer lowercases.
+    ///
+    /// Defaults to `true`.
+    pub fn with_lowercase(mut self, lowercase: bool) -> Self {
+        self.lowercase = lowercase;
+        self
+    }
+
+    /// Sets the token size for the tokenizer and the model.
+    ///
+    /// Defaults to [`BertModel::TOKEN_RANGE`].
```

Comment:
Here it was ``[`K::TOKEN_RANGE`]``, I put `BertModel` since it is the bound on `K`. Maybe we can find a better way to fix this.